File: 01-tts-callout/scripts/callout.py
# Sinch Voice Callout — dials a phone number, plays an audio file, then a TTS message.
# Requirements: pip install requests python-dotenv

import os
import sys
import requests
from dotenv import load_dotenv

# Load credentials from project root .env
load_dotenv(dotenv_path=os.path.join(os.path.dirname(__file__), "../../.env"))

project_id         = os.environ.get("PROJECT_ID")
key_id             = os.environ.get("KEY_ID")
key_secret         = os.environ.get("KEY_SECRET")
sinch_number       = os.environ.get("SINCH_NUMBER")
destination_number = os.environ.get("DESTINATION_NUMBER")

for var, name in [
    (project_id, "PROJECT_ID"),
    (key_id, "KEY_ID"),
    (key_secret, "KEY_SECRET"),
    (sinch_number, "SINCH_NUMBER"),
    (destination_number, "DESTINATION_NUMBER"),
]:
    if not var:
        print(f"ERROR: {name} is not set. Copy .env.example to .env and fill in your credentials.", file=sys.stderr)
        sys.exit(1)

base_url = f"https://voice.api.sinch.com/v2/projects/{project_id}/calls"

# SVAML payload: dial -> on answer play audio file -> hangup
payload = {
    "commands": [
        {
            "command": "dial",
            "name": "audio-notification",
            "from": {
                "type": "PHONE",
                "phone": {"number": sinch_number}
            },
            "to": {
                "type": "PHONE",
                "phone": {"number": destination_number}
            },
            "dialTimeout": "30s",
            "maxDuration": "5m",
            "events": {
                "onAnswer": [
                    {
                        "command": "messages",
                        "name": "notification",
                        "messages": [
                            {
                                "type": "PLAY",
                                "play": {
                                    "url": "https://samplelib.com/mp3/sample-12s.mp3"
                                }
                            },
                            {
                                "type": "SAY",
                                "say": {
                                    "text": "Hello! This is a test notification from Sinch. Your verification code is 4 8 3 7.",
                                    "voiceName": "Emma"
                                }
                            }
                        ],
                        "events": {
                            "onFinish": [
                                {"command": "hangup"}
                            ]
                        }
                    }
                ]
            }
        }
    ]
}

print(f"Placing callout from {sinch_number} to {destination_number} ...")

try:
    response = requests.post(
        base_url,
        json=payload,
        auth=(key_id, key_secret)
    )
    data = response.json()

    if response.status_code == 201:
        print("Call created successfully:")
        import json
        print(json.dumps(data, indent=2))
    else:
        print(f"ERROR {response.status_code}:", file=sys.stderr)
        import json
        print(json.dumps(data, indent=2), file=sys.stderr)
        sys.exit(1)

except requests.RequestException as e:
    print(f"Request failed: {e}", file=sys.stderr)
    sys.exit(1)

File: 01-tts-callout/scripts/callout.java
// Sinch Voice Callout - dials a phone number, plays an audio file, then a TTS message.
// Requires Java 11+ (uses java.net.http.HttpClient).
//
// Maven pom.xml dependencies: none beyond JDK 11+
// Compile: javac -d out Callout.java
// Run:     java -cp out com.sinch.tutorials.ttscallout.Callout
//
// Variables are loaded from ../../.env (relative to the working directory).
// If the .env file is missing, system environment variables are used as a fallback.
//
// .env file format (one per line):
//   PROJECT_ID=your_project_id
//   KEY_ID=your_key_id
//   KEY_SECRET=your_key_secret
//   SINCH_NUMBER=+1234567890
//   DESTINATION_NUMBER=+0987654321

package com.sinch.tutorials.ttscallout;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Base64;
import java.util.HashMap;
import java.util.Map;

public class Callout {

    private static final Map<String, String> dotEnv = new HashMap<>();

    public static void main(String[] args) throws Exception {
        // Load .env file from two directories up (../../.env)
        loadDotEnv(Path.of("../../.env"));

        String projectId         = requireEnv("PROJECT_ID");
        String keyId             = requireEnv("KEY_ID");
        String keySecret         = requireEnv("KEY_SECRET");
        String sinchNumber       = requireEnv("SINCH_NUMBER");
        String destinationNumber = requireEnv("DESTINATION_NUMBER");

        String baseUrl = "https://voice.api.sinch.com/v2/projects/" + projectId + "/calls";

        // Basic Auth header: base64("keyId:keySecret")
        String credentials = Base64.getEncoder()
                .encodeToString((keyId + ":" + keySecret).getBytes());

        // SVAML payload: dial -> on answer play audio file -> hangup
        String body = String.format("""
            {
              "commands": [
                {
                  "command": "dial",
                  "name": "audio-notification",
                  "from": {
                    "type": "PHONE",
                    "phone": { "number": "%s" }
                  },
                  "to": {
                    "type": "PHONE",
                    "phone": { "number": "%s" }
                  },
                  "dialTimeout": "30s",
                  "maxDuration": "5m",
                  "events": {
                    "onAnswer": [
                      {
                        "command": "messages",
                        "name": "notification",
                        "messages": [
                          {
                            "type": "PLAY",
                            "play": {
                              "url": "https://samplelib.com/mp3/sample-12s.mp3"
                            }
                          },
                          {
                            "type": "SAY",
                            "say": {
                              "text": "Hello! This is a test notification from Sinch. Your verification code is 4 8 3 7.",
                              "voiceName": "Emma"
                            }
                          }
                        ],
                        "events": {
                          "onFinish": [
                            { "command": "hangup" }
                          ]
                        }
                      }
                    ]
                  }
                }
              ]
            }
            """, sinchNumber, destinationNumber);

        System.out.println("Placing callout from " + sinchNumber + " to " + destinationNumber + " ...");

        HttpClient client = HttpClient.newHttpClient();
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(baseUrl))
                .header("Content-Type", "application/json")
                .header("Authorization", "Basic " + credentials)
                .POST(HttpRequest.BodyPublishers.ofString(body))
                .build();

        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
        int statusCode = response.statusCode();
        String responseBody = response.body();

        if (statusCode == 201) {
            System.out.println("Call created successfully:");
            System.out.println(responseBody);
        } else {
            System.err.println("ERROR " + statusCode + ":");
            System.err.println(responseBody);
            System.exit(1);
        }
    }

    /**
     * Loads key=value pairs from a .env file into the dotEnv map.
     * Blank lines and lines starting with '#' are ignored.
     * Surrounding quotes on values are stripped.
     */
    private static void loadDotEnv(Path path) {
        Path resolved = path.toAbsolutePath().normalize();
        if (!Files.exists(resolved)) {
            System.out.println("Note: .env file not found at " + resolved + ", falling back to system environment variables.");
            return;
        }
        try {
            for (String line : Files.readAllLines(resolved)) {
                line = line.trim();
                if (line.isEmpty() || line.startsWith("#")) continue;
                int eqIndex = line.indexOf('=');
                if (eqIndex < 0) continue;
                String key = line.substring(0, eqIndex).trim();
                String value = line.substring(eqIndex + 1).trim();
                // Strip surrounding quotes if present
                if (value.length() >= 2
                        && ((value.startsWith("\"") && value.endsWith("\""))
                        ||  (value.startsWith("'")  && value.endsWith("'")))) {
                    value = value.substring(1, value.length() - 1);
                }
                dotEnv.put(key, value);
            }
            System.out.println("Loaded " + dotEnv.size() + " variable(s) from " + resolved);
        } catch (IOException e) {
            System.err.println("Warning: could not read .env file: " + e.getMessage());
        }
    }

    /**
     * Returns the value for a variable, checking the .env map first,
     * then falling back to system environment variables.
     */
    private static String requireEnv(String name) {
        String value = dotEnv.get(name);
        if (value == null || value.isBlank()) {
            value = System.getenv(name);
        }
        if (value == null || value.isBlank()) {
            System.err.println("ERROR: " + name + " is not set in .env or system environment.");
            System.exit(1);
        }
        return value;
    }
}